packages/orm/src/FlushProxy.ts

```diff
@@ -0,0 +1,120 @@
+import { OneToManyCollection } from "./collections/OneToManyCollection";
+import { ManyToManyCollection } from "./collections/ManyToManyCollection";
+import { ManyToOneReference } from "./collections/ManyToOneReference";
+import { Entity, EntityManager } from "./EntityManager";
+import * as util from "util";
+
+// These proxies are passed into hooks during a flush to enable changes mid flush
+
+type FlushProxy<T> = T & { __flushSecret: number; __proxyTarget: T };
+type Loader = OneToManyCollection<any, any> | ManyToManyCollection<any, any> | ManyToOneReference<any, any, any>;
+
+export function isFlushProxy<T>(objectOrProxy: T): objectOrProxy is FlushProxy<T> {
+  return util.types.isProxy(objectOrProxy) && (objectOrProxy as any).__flushSecret !== undefined;
+}
+
+export function getTargetFromMaybeFlushProxy<T>(objectOrProxy: T | FlushProxy<T>): T {
+  return isFlushProxy(objectOrProxy) ? objectOrProxy.__proxyTarget : objectOrProxy;
+}
+
+function proxyLoaderResult(result: any, flushSecret: number) {
+  if (Array.isArray(result)) {
+    return result.map((entity) => proxyEntity(entity, flushSecret));
+  } else if (typeof result === "object") {
+    return proxyEntity(result, flushSecret);
+  }
+
+  return result;
+}
+
+function createFlushProxy<T extends object>(
+  target: T,
+  flushSecret: number,
+  block: (target: T, property: PropertyKey, receiver: any) => any,
+): FlushProxy<T> {
+  if (isFlushProxy(target)) {
+    return target;
+  }
+
+  return new Proxy(target, {
+    get(target, property, receiver) {
+      if (property === "__flushSecret") {
+        return flushSecret;
+      }
+
+      if (property === "__proxyTarget") {
```

Comment:
I'm cool with this trap, and I've done it several times before, just copy/pasting that I also saw this way:

```
const proxies = new WeakSet();

export function createProxy(obj) {
    const handler = {};
    const proxy = new Proxy(obj, handler);
    proxies.add(proxy);
    return proxy;
}

export function isProxy(obj) {
    return proxies.has(obj);
}
```

Or doing proxy detection and de-wrapping that seems fine too.